### PR TITLE
Support setting ballast size in percentage of total Mem in ballast extension

### PR DESCRIPTION
**Pull request by mxiamxia**
**Descritpion:** 
Enhance the current ballast extension which is to set memory ballast for the collector process.  And we'll remove `--mem-ballast-size-mib` support from command line later.

**The changes,**
Added new configuration `size_in_percentage` which can dynamically set ballast size based on the total memory of the running environments. (hosts, Docker, K8s, etc)

**Sample Config,**
```yaml
extensions:
  memory_ballast:
    size_in_percentage: 20
```

**How ballast size is calculated by percentage setup**
When `size_in_percentage` is enabled with the value(1-100), the `ballast_size` will be calculated by 
`size_in_percentage * totalMemory`. The `totalMemory` will be calculated for hosts and containers(in docker, k8s, etc) by the following steps,
1. Look up Memory Cgroup subsystem where the collector is running on(it could be container, VM or bare metal host), check if there is total memory limitation has been set for the collector process.
The total assigned memory for the process is in `memory.limit_in_bytes` file under cgroup (eg, `/sys/fs/cgroup/memory/memory.limit_in_bytes`).
   
2. If `memory.limit_in_bytes` has a positive value other than `9223372036854771712`. The `ballest_size`
   will be calculated by `memory.limit_in_bytes * size_in_percentage`. Else If `memory.limit_in_bytes` value is `9223372036854771712`(`0x7FFFFFFFFFFFF000`), it indicates there is no memory limit has been set for the collector process in cgroup. Then the `totalMemoery` will be determined in the next step
3. if there is no memory limit set in cgroup for the collector process or container where the collector is running. The total memory will be calculated by `github.com/shirou/gopsutil/mem` which is supported in multiple OS systems.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector/issues/2516

**Test:** 
Tested locally on EC2, Docker and K8S for `ballast_size_in_percentage` configuration

### CODE REVIEW

**Comment by alolita** [comment]
@mxiamxia can you fix these failing tests. Thx.

File: extension/ballastextension/config_test.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ func TestLoadConfig(t *testing.T) {
 		&Config{
 			ExtensionSettings: config.NewExtensionSettings(config.NewIDWithName(typeStr, "1")),
 			SizeMiB:           123,
+			SizeInPercentage:  20,
```
Comment (bogdandrutu) [inline extension/ballastextension/config_test.go:48]: I think we had a problem with this, was supported only on some platforms/environments. Is that still the case? If yes then we need to document it. Maybe I miss remember, but please double check.

File: service/defaultcomponents/defaults.go [hunk 2]
```diff
@@ -63,6 +64,7 @@ func Components() (
 		oidcauthextension.NewFactory(),
 		pprofextension.NewFactory(),
 		zpagesextension.NewFactory(),
+		ballastextension.NewFactory(),
```
Comment (bogdandrutu) [inline service/defaultcomponents/defaults.go:67]: Please split this in a separate PR, we need to discuss about deprecating the flag in that PR, also the flag is accessed by the memory limiter processor so we need a solution for that as well.

Comment (mxiamxia) [reply service/defaultcomponents/defaults.go]: Sure. 
> the flag is accessed by the memory limiter processor so we need a solution for that as well.

For this one, I am thinking to expose the final `ballast_size` from the extension and force the memory_limiter to use the value set from ballast extension. we can discuss it later.

```
// GetBallastSize allows the external to get the current ballast size setting
func GetBallastSize() uint64 {
	return ballastSizeBytes
}
```

File: extension/ballastextension/config_test.go (hunk 1 above)
Comment (mxiamxia) [reply extension/ballastextension/config_test.go:48]: I tested it on mac, windows, linux, docker and k8s and no issue found. 

To get `total memory` used for the percentage calculation first by `cgroup`(for linux*), if cgroup is not set or not supported then we'll get it by `gopsutils` which is cross-platform supported.

**Comment by punya** [comment]
@Aneurysm9 @anuraaga could you please take a look?

File: extension/ballastextension/memory_ballast.go [hunk 3]
```diff
@@ -25,17 +25,32 @@ import (
...
+	// absolute value supersedes percentage setting
 	if m.cfg.SizeMiB > 0 {
-		ballastSizeBytes := uint64(m.cfg.SizeMiB) * megaBytes
+		ballastSizeBytes = m.cfg.SizeMiB * megaBytes
+	} else {
+		totalMemory, err := m.memGetter()
+		if err != nil {
+			return err
+		}
+		ballastPercentage := m.cfg.SizeInPercentage
```
Comment (anuraaga) [inline extension/ballastextension/memory_ballast.go:44]: I might miss it but is the config value validated?

File: internal/iruntime/total_memory_linux.go [hunk 4]
```diff
@@ -18,9 +18,13 @@ package iruntime
 
 import "go.opentelemetry.io/collector/internal/cgroups"
 
+// _unlimitMemorySize defines the bytes size when memory limit is not set
+// for the container and process with cgroups
+const _unlimitMemorySize = 9223372036854771712
```
Comment (anuraaga) [inline internal/iruntime/total_memory_linux.go:23]: unlimited

File: extension/ballastextension/memory_ballast.go [hunk 5]
```diff
@@ -25,17 +25,32 @@ import (
 const megaBytes = 1024 * 1024
 
 type memoryBallast struct {
-	cfg     *Config
-	logger  *zap.Logger
-	ballast []byte
+	cfg       *Config
+	logger    *zap.Logger
+	ballast   []byte
+	memGetter func() (uint64, error)
```
Comment (anuraaga) [inline extension/ballastextension/memory_ballast.go:31]: I think it makes sense to name this sort of func as a func name unless there's some go style I don't know about. E.g. `getTotalMem`

**Review by anuraaga — COMMENTED** [review]
Cool feature

File: internal/iruntime/total_memory_linux.go (hunk 4 above)
Comment (bogdandrutu) [reply internal/iruntime/total_memory_linux.go:23]: Why "_" as prefix?

Comment (mxiamxia) [reply internal/iruntime/total_memory_linux.go:23]: "_" is naming convention to indicate the variables for internal use. I guess I'm fine to remove it  since it is restricted by lowercase in go.

File: extension/ballastextension/memory_ballast.go (hunk 3 above)
Comment (mxiamxia) [reply extension/ballastextension/memory_ballast.go:44]: Added `config.Validate()`

File: extension/ballastextension/README.md [hunk 6]
```diff
@@ -6,12 +6,38 @@ Memory Ballast extension enables applications to configure memory ballast for th
...
+1. Look up Memory Cgroup subsystem on the target host or container, find out if there is any total memory limitation has been set for the running collector process.
+   Check the value in `memory.limit_in_bytes` file under cgroup memory files (eg, `/sys/fs/cgroup/memory/memory.limit_in_bytes`).
+
+2. If `memory.limit_in_bytes` is positive value other than `9223372036854771712`. The `ballest_size`
+   will be calculated by `memory.limit_in_bytes * size_in_percentage`.
+   If `memory.limit_in_bytes` value is `9223372036854771712`(`0x7FFFFFFFFFFFF000`), it indicates there is no memory limit has
+   been set for the collector process or the running container in cgroup. Then the `totalMemoery` will be determined in next step.
+   
+3. if there is no memory limit set in cgroup for the collector process or container where the collector is running. The total memory will be
+   calculated by `github.com/shirou/gopsutil/mem` which is supported in multiple OS systems.
```
[the remark this answers is not in the document]
Comment (mxiamxia) [reply extension/ballastextension/README.md]: updated. Thanks!

**Comment by alolita** [comment]
@anuraaga @Aneurysm9 can you please re-review? The tests are now passing.

File: extension/ballastextension/README.md [hunk 7]
```diff
@@ -6,12 +6,38 @@ Memory Ballast extension enables applications to configure memory ballast for th
...
+   If `memory.limit_in_bytes` value is `9223372036854771712`(`0x7FFFFFFFFFFFF000`), it indicates there is no memory limit has
+   been set for the collector process or the running container in cgroup. Then the `totalMemory` will be determined in next step.
+   
+3. if there is no memory limit set in cgroup for the collector process or container where the collector is running. The total memory will be
+   calculated by `github.com/shirou/gopsutil/mem` on `mem.VirtualMemory().total` which is supported in multiple OS systems.
 
-Example:
 
+Example:
+Config memory ballast size by absolut value in Mib
```
Comment (Aneurysm9) [inline extension/ballastextension/README.md:31]:
```suggestion
Config memory ballast size by absolute value in Mib
```

File: internal/iruntime/total_memory_other.go [hunk 8]
```diff
@@ -16,14 +16,8 @@
...
 package iruntime
 
-import (
-	"fmt"
-)
-
-var errTotalMemoryNotAvailable = fmt.Errorf("reading cgroups total memory is available only on linux")
-
 // TotalMemory returns total available memory.
 // This is non-Linux version that returns -1 and errTotalMemoryNotAvailable.
```
Comment (Aneurysm9) [inline internal/iruntime/total_memory_other.go:20]: This returns a `uint64` so it can't return `-1`.  It also doesn't return `errTotalMemoryNotAvailable`.

Comment (mxiamxia) [reply internal/iruntime/total_memory_other.go]: Thanks for catching it.

**Review by Aneurysm9 — APPROVED** [review]

File: extension/ballastextension/README.md [hunk 9]
```diff
@@ -6,12 +6,38 @@ Memory Ballast extension enables applications to configure memory ballast for th
...
 
+Example:
+Config memory ballast size by absolute value in Mib
 ```yaml
 extensions:
   memory_ballast:
     size_mib: 64
 ```
+
+Config memory ballast size in percentage of total memory
```
Comment (rakyll) [inline extension/ballastextension/README.md:38]: Ditto

Config that uses 20% of the total memory for the ballast:

File: extension/ballastextension/README.md [hunk 10]
```diff
@@ -6,12 +6,38 @@ Memory Ballast extension enables applications to configure memory ballast for th
...
+   If `memory.limit_in_bytes` value is `9223372036854771712`(`0x7FFFFFFFFFFFF000`), it indicates there is no memory limit has
+   been set for the collector process or the running container in cgroup. Then the `totalMemory` will be determined in next step.
+   
+3. if there is no memory limit set in cgroup for the collector process or container where the collector is running. The total memory will be
+   calculated by `github.com/shirou/gopsutil/mem` on `mem.VirtualMemory().total` which is supported in multiple OS systems.
 
-Example:
 
+Example:
+Config memory ballast size by absolute value in Mib
```
Comment (rakyll) [inline extension/ballastextension/README.md:31]: Maybe say something like

Config that uses 64 Mib of memory for the ballast:

**Review by rakyll — CHANGES_REQUESTED** [review]

File: extension/ballastextension/README.md [hunk 11]
```diff
@@ -6,12 +6,38 @@ Memory Ballast extension enables applications to configure memory ballast for th
...
+`size_in_percentage * totalMemory`. The `totalMemory` can be retrieved for hosts and containers(in docker, k8s, etc) by the following steps,
+1. Look up Memory Cgroup subsystem on the target host or container, find out if there is any total memory limitation has been set for the running collector process.
+   Check the value in `memory.limit_in_bytes` file under cgroup memory files (eg, `/sys/fs/cgroup/memory/memory.limit_in_bytes`).
+
+2. If `memory.limit_in_bytes` is positive value other than `9223372036854771712`(`0x7FFFFFFFFFFFF000`). The `ballest_size`
+   will be calculated by `memory.limit_in_bytes * size_in_percentage`.
+   If `memory.limit_in_bytes` value is `9223372036854771712`(`0x7FFFFFFFFFFFF000`), it indicates there is no memory limit has
+   been set for the collector process or the running container in cgroup. Then the `totalMemory` will be determined in next step.
+   
+3. if there is no memory limit set in cgroup for the collector process or container where the collector is running. The total memory will be
```
[the remark this answers is not in the document]
Comment (mxiamxia) [reply extension/ballastextension/README.md:26]: Added `gopsutil` github link

File: extension/ballastextension/README.md (hunk 10 above)
Comment (mxiamxia) [reply extension/ballastextension/README.md]: Yep, updated both.

**Comment by mxiamxia** [comment]
Hi @anuraaga @bogdandrutu @tigrannajaryan @rakyll , PTAL! Thanks.

**Review by anuraaga — APPROVED** [review]

**Review by bogdandrutu — APPROVED** [review]
@tigrannajaryan @rakyll PTAL, will merge tomorrow otherwise

**Review by tigrannajaryan — COMMENTED** [review]
LGTM, except I have just found we also have a built-in ballast implementation in https://github.com/open-telemetry/opentelemetry-collector/blob/6cafee91d2c3f2151f46275691bd8c1607ea1ab9/service/collector.go#L317

I am confused. Why do we have 2 implementations? Are they related or competing?

**Comment by mxiamxia** [comment]
> opentelemetry-collector/service/collector.go

This one is taking the `ballast size` input from the command line. I will send another PR to deprecate the old ballast memory creation code.

**Review by rakyll — APPROVED** [review]

**Comment by alolita** [comment]
@bogdandrutu this PR is ready to be merged. :-)
